packages/grafana-data/src/transformations/transformDataFrame.ts

```diff
@@ -1,37 +1,71 @@
+
 import { DataFrame, DataTransformerConfig } from '../types';
 import { standardTransformersRegistry } from './standardTransformersRegistry';
 
+const getTransform = (config: DataTransformerConfig, data: Observable<DataFrame[]>): Observable<DataFrame[]> => {
+  const info = standardTransformersRegistry.get(config.id);
+
+  if (!info) {
+    return data;
+  }
+
+  return data.pipe(
+    mergeMap(before => {
+      const defaultOptions = info.transformation.defaultOptions ?? {};
+      const options = { ...defaultOptions, ...config.options };
+
+      return info.transformation.transformer(options, before).pipe(
+        map(after => {
+          if (after === before) {
+            return after;
+          }
+
+          // Add a key to the metadata if the data changed
+          for (const series of after) {
+            if (!series.meta) {
+              series.meta = {};
+            }
+
+            if (!series.meta.transformations) {
+              series.meta.transformations = [info.id];
+            } else {
+              series.meta.transformations = [...series.meta.transformations, info.id];
+            }
+          }
+
+          return after;
+        })
+      );
+    })
+  );
+};
+
 /**
  * Apply configured transformations to the input data
  */
-export function transformDataFrame(options: DataTransformerConfig[], data: DataFrame[]): DataFrame[] {
-  let processed = data;
-  for (const config of options) {
-    const info = standardTransformersRegistry.get(config.id);
-
-    if (!info) {
-      return data;
-    }
-
-    const defaultOptions = info.transformation.defaultOptions ?? {};
-    const options = { ...defaultOptions, ...config.options };
-    const transformer = info.transformation.transformer(options);
-    const after = transformer(processed);
-
-    // Add a key to the metadata if the data changed
-    if (after && after !== processed) {
-      for (const series of after) {
-        if (!series.meta) {
-          series.meta = {};
-        }
-        if (!series.meta.transformations) {
-          series.meta.transformations = [info.id];
-        } else {
-          series.meta.transformations = [...series.meta.transformations, info.id];
-        }
-      }
-      processed = after;
-    }
+export function transformDataFrame(options: DataTransformerConfig[], data: DataFrame[]): Observable<DataFrame[]> {
+  if (!options.length) {
+    return of(data);
   }
-  return processed;
+
+  return of(data).pipe(
+    mergeMap(source => {
+      const observables: Array<Observable<DataFrame[]>> = [];
+
+      for (let index = 0; index < options.length; index++) {
+        const config = options[index];
+        const prev = index === 0 ? of(source) : observables[index - 1];
+        const curr = prev.pipe(
+          mergeMap(d => {
+            return getTransform(config, of(d));
+          })
+        );
+        observables.push(curr);
+      }
+
+      return observables[observables.length - 1];
+    })
```

Comment:
```suggestion
 const operators: Array<OperatorFunction<DataFrame[], DataFrame[]>> = mapConfigToOperators(options);
 const observable = of(data);

 return observable.pipe.apply(observable, operators); // we might need to run each of this befor applying them. It is just pseudo code.
```

hmm, I have not tested this code but what I am thinking is that each transform should just return an `OperatorFunction<DataFrame[], DataFrame[]>`. Then we create a stream out of the incoming data and pipe that stream through all the operators and return the result.

`mapConfigToOperators` should probably wrap each operator in another operator function that applies the meta data.

Reply:
I really like this approach, very clever using the apply function. Never thought of using that before, great ⭐

Reply:
Unfortunately TypeScript couldn't understand the typings as `apply` returns `any` but I left it as is with a `// @ts-ignore`